File: apps/shared-ts/src/ids.ts
export type UserID = `usr_${string}`
export type SessionID = `sess_${string}`
export type ProjectID = `prj_${string}`

const isUserID = (id: string): id is UserID => id.startsWith("usr_")

export const isSessionID = (id: unknown): id is SessionID =>
  typeof id === "string" && id.startsWith("sess_")

export const isProjectID = (id: unknown): id is ProjectID =>
  typeof id === "string" && id.startsWith("prj_")

export const userIDFromString = (id: string): UserID | undefined => {
  if (!isUserID(id)) {
    return undefined
  }

  return id
}

export const sessionIDFromString = (id: string): SessionID | undefined => {
  if (!isSessionID(id)) {
    return undefined
  }

  return id
}

export const projectIDFromString = (id: string): ProjectID | undefined => {
  if (!isProjectID(id)) {
    return undefined
  }

  return id
}

File: apps/api/internal/domain/ids.go
package domain

import "strings"

type UserID string

type SessionID string

type ProjectID string

type ExpenseID string

func UserIDFromString(id string) (UserID, bool) {
	if !strings.HasPrefix(id, "usr_") {
		return "", false
	}
	return UserID(id), true
}

func (id UserID) String() string {
	return string(id)
}

func SessionIDFromString(id string) (SessionID, bool) {
	if !strings.HasPrefix(id, "sess_") {
		return "", false
	}
	return SessionID(id), true
}

func (id SessionID) String() string {
	return string(id)
}

func ProjectIDFromString(id string) (ProjectID, bool) {
	if !strings.HasPrefix(id, "prj_") {
		return "", false
	}
	return ProjectID(id), true
}

func (id ProjectID) String() string {
	return string(id)
}

func ExpenseIDFromString(id string) (ExpenseID, bool) {
	if !strings.HasPrefix(id, "exp_") {
		return "", false
	}
	return ExpenseID(id), true
}

func (id ExpenseID) String() string {
	return string(id)
}
